acme/acme/messages.py

```diff
@@ -274,19 +296,27 @@ class Registration(ResourceBody):
     agreement = jose.Field('agreement', omitempty=True)
     status = jose.Field('status', omitempty=True)
     terms_of_service_agreed = jose.Field('termsOfServiceAgreed', omitempty=True)
+    external_account_binding = jose.Field('externalAccountBinding', omitempty=True)
 
     phone_prefix = 'tel:'
     email_prefix = 'mailto:'
 
     @classmethod
-    def from_data(cls, phone=None, email=None, **kwargs):
+    def from_data(cls, account_public_key=None, kid=None,
+                  hmac_key=None, phone=None, email=None,
+                  directory=None, **kwargs):
         """Create registration resource from contact details."""
         details = list(kwargs.pop('contact', ()))
         if phone is not None:
             details.append(cls.phone_prefix + phone)
         if email is not None:
             details.append(cls.email_prefix + email)
         kwargs['contact'] = tuple(details)
+
+        if kid is not None and hmac_key is not None:
```

Comment:
You're totally right about the directory having to be passed in to construct the object. Since the `ExternalAccountBinding` object needs so many parameters that are only for it, let's move the `from_data` call into `client.py` so that we only have to pass one object that might be `None`, instead of four separate things. Then we could clean up that `from_data` call signature.

Reply:
Just pushed a change to PR: https://github.com/certbot/certbot/pull/6059/commits/8963fae5dcf99b96c65c2d3d5ea7b1f9a403138a

This what you meant?

Reply:
Didn't do cleanup of old stuff. But will do it if this is what we're after.

Reply:
Should have been more specific, when I said `client.py` above I mean `certbot/certbot/client.py`. The paradigm we have here is mostly constructing objects in the client code or using `from_data` methods, rather than a `build_*` paradigm. So in `certbot/certbot/client.py:perform_registration`, that would look something like:

```
if conditions_checking_if_appropriate_to_construct_eab:
    eab = ExternalAccountBinding.from_data(account_public_key, kid, hmac_key, directory)
else:
    eab = None
try:
    acme.new_account_and_tos(
        messages.NewRegistration.from_data(email=config.email, external_account_binding=eab),
        tos_cb)
```

Reply:
Aah, will fix!

Reply:
New version pushed.